lru.h

```diff
@@ -120,24 +121,34 @@ template<typename F> struct ConcurrentLru {
             typename boost::remove_reference<
                 typename F::result_type>::type
             >::type const;
-        using result_type = std::shared_ptr<underlying_type>;
+        using result_type = std::shared_future<std::shared_ptr<underlying_type>>;
+
         result_type operator()(argument_type arg) const {
-            return std::make_shared<underlying_type>(f(arg));
+            //build a future that will be lazy initialized
+            return std::async(std::launch::deferred, [&](){
+                return std::make_shared<underlying_type>(f(arg));
+            }).share();
         }
     };
     Lru<SharedPtrF> lru;
     std::unique_ptr<std::mutex> mutex = std::make_unique<std::mutex>();
 
 public:
-    using result_type = typename SharedPtrF::result_type;
+    using result_type = typename std::shared_ptr<typename SharedPtrF::underlying_type>;
     using argument_type = typename SharedPtrF::argument_type;
 
     ConcurrentLru(F fun, size_t max = 10): lru(SharedPtrF{std::move(fun)}, max) {}
     ConcurrentLru(ConcurrentLru&&) = default;// needed by old version of gcc
 
     result_type operator()(argument_type arg) const {
-        std::lock_guard<std::mutex> lock(*mutex);
-        return lru(arg);
+        typename SharedPtrF::result_type future;
+        {
+            std::lock_guard<std::mutex> lock(*mutex);
```

Comment:
I don't think the mutex needs to be a `std::unique_ptr<std::mutex>`. It's private and never exposed. A `std::mutex` should be enought.

Reply:
I agree that this is not part of the scope of this PR, therefor I'm fine with it.

Reply:
I agree with you, it seemed strange but I didn't want to change that at the same time